### PR TITLE
Make BigQueryRow.GetFieldIndex public

### PR DESCRIPTION
**Pull request by roji**
This API can be useful in general, and can be relied upon by the ADO.NET provider.

### CODE REVIEW

**Comment by jskeet** [comment]
Rather than making this public here, how would you feel about making it an extension method on `TableSchema`? The schema itself is already public.

**Comment by roji** [comment]
It's definitely possible. Do imagine the future pre-processed schema as being of a completely different type as the current `TableSchema`? If so then it might be a good idea to provide this API on the row to avoid breakage as you suggest. We could also provide both, with the row API simply delegating to the schema one for now.

One question - is it a design principle to implement as extensions anything that *can* be implemented as an extension (i.e. doesn't access private/internals)? Am interested as to the reason for this...

**Comment by jskeet** [comment]
Yes, the pre-processed schema would have a reference to the existing `TableSchema`, but probably additional information too. `TableSchema` is defined in the generated REST assembly, so we can't change that, basically.

In terms of "implement everything possible as extension methods" - nope, I just like to keep the public API surface as orthogonal as possible. While the method is private, there's no benefit in it being an extension method - but once it's public, it really feels like a more schema-like concern than a row-like concern.

**Comment by roji** [comment]
Ah OK, I didn't realize `TableSchema` is out of our hands.

Do you see the pre-processed schema as ending up as a subclass of `TableSchema`, simply adding some things? If so, the extension method approach seems like it would still work (at worst it could check if the schema it's running on is pre-processed or not).

For now, I amended the PR to refactor GetFieldIndex as an extension to TableSchema, and removed it from BigQueryRow.

**Comment by jskeet** [comment]
No, I don't anticipate making it a subclass of `TableSchema` - I'd favour composition over inheritance, as I generally do. (Making it a subclass makes it hard to create a new instance of it from an existing `TableSchema` - we'd need to copy over all the state, some of which we may not even have access to.)

Let's go with the extension method for now - will rereview the change.

File: apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/TableSchemaExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    /// Extension methods over table schemas.
+    /// </summary>
+    public static class TableSchemaExtensions
+    {
+        /// <summary>
+        /// Retrieves the index of a field given its name.
+        /// </summary>
+        public static int GetFieldIndex(this TableSchema schema, string fieldName)
+        {
+            for (int i = 0; i < schema.Fields.Count; i++)
```
Comment (jskeet) [inline apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/TableSchemaExtensions.cs]: Now that this is in the public API, let's add preconditions. (Arguably they should have been there before for the name...)

```csharp
GaxPreconditions.CheckNotNull(schema, nameof(schema);
GaxPreconditions.CheckNotNull(fieldName, nameof(fieldName);
```

Comment (roji) [inline apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/TableSchemaExtensions.cs]: Done

**Comment by jskeet** [comment]
Great, thanks - will merge when green.

**Pull request merged**